[JIT] add support for torch.jit.Final in python 3.6

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47393 [JIT] add support for torch.jit.Final in python 3.6**

Python added support for typing.Final in 3.8, and there exists a backport for 3.7 and 3.6 in typing_extensions.Final

The 3.6 version of typing_extensions.Final doesn't support the `getattr(x, '__origin__')` to determine if an object is of the Final type while 3.7 does. This PR allows for the Final type of an object to be discoverable in 3.6


Differential Revision: [D24739402](https://our.internmc.facebook.com/intern/diff/D24739402)

Commit:
[JIT] add support for torch.jit.Final in python 3.6

[ghstack-poisoned]

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -717,7 +717,7 @@ def is_rref(ann):
 def is_final(ann):
     return ann.__module__ in {'typing', 'typing_extensions'} and \
-        (getattr(ann, '__origin__', None) is Final)
+        (getattr(ann, '__origin__', None) is Final or isinstance(ann, type(Final)))
```
Optional:

Can you say a bit more about why this change is needed (in PR description)? I am curious.

Lilyjjo replied:
added :)

SplitInfinity commented:
Do we need a test for this? I think the issue that reported this issue has a small test case we can use.

Commit:
Update on "[JIT] add support for torch.jit.Final in python 3.6"


Differential Revision: [D24739402](https://our.internmc.facebook.com/intern/diff/D24739402)

[ghstack-poisoned]

Commit:
Update on "[JIT] add support for torch.jit.Final in python 3.6"


Python added support for typing.Final in 3.8, and there exists a backport for 3.7 and 3.6 in typing_extensions.Final

The 3.6 version of typing_extensions.Final doesn't support the getattr(x, '__origin__') to determine if an object is of the Final type while 3.7 does. This PR allows for the Final type of an object to be discoverable in 3.6


Differential Revision: [D24739402](https://our.internmc.facebook.com/intern/diff/D24739402)

[ghstack-poisoned]

Lilyjjo commented on test/jit/test_recursive_script.py:
```diff
@@ -159,7 +158,6 @@ def a(x):
             # Make sure that no entries are left over from the previous failure
             FileCheck().check_count("is being compiled", 2).run(str(e))
 
-    @unittest.skipIf(sys.version_info[:2] < (3, 7), "Class annotations are a thing in > 3.5, need to fix for < 3.7")
```
I think this comment was out of date? Not totally sure, but this test now passes for python3.6 and python3.7 and checks the functionality of Final

Commit:
Update on "[JIT] add support for torch.jit.Final in python 3.6"


Python added support for typing.Final in 3.8, and there exists a backport for 3.7 and 3.6 in typing_extensions.Final

The 3.6 version of typing_extensions.Final doesn't support the `getattr(x, '__origin__')` to determine if an object is of the Final type while 3.7 does. This PR allows for the Final type of an object to be discoverable in 3.6


Differential Revision: [D24739402](https://our.internmc.facebook.com/intern/diff/D24739402)

[ghstack-poisoned]

Commit:
Update on "[JIT] add support for torch.jit.Final in python 3.6"


Python added support for typing.Final in 3.8, and there exists a backport for 3.7 and 3.6 in typing_extensions.Final

The 3.6 version of typing_extensions.Final doesn't support the `getattr(x, '__origin__')` to determine if an object is of the Final type while 3.7 does. This PR allows for the Final type of an object to be discoverable in 3.6


Differential Revision: [D24739402](https://our.internmc.facebook.com/intern/diff/D24739402)

[ghstack-poisoned]
